### PR TITLE
First pass at gRPC transcoding improvements

### PR DESCRIPTION
**Pull request by jskeet**
This isn't yet absolutely complete, but:

- The refactoring work is about right now, I think
- It introduces the data-file-driven tests which I think we'll want to adopt across languages eventually

I'd like to get it in as the basis for other REGAPIC work.

### CODE REVIEW

**Comment by jskeet** [comment]
Definitely best to review one commit at a time.

File: Google.Api.Gax.Grpc/Rest/RestMethod.cs [hunk 1]
```diff
@@ -9,198 +9,65 @@
...
-            {
-                throw new RpcException(status, httpResponse.GetTrailers());
-            }
-            return (TResponse) _parser.Parse(httpResponse.Content, _protoMethod.OutputType);
-        }
+    internal HttpRequestMessage CreateRequest(IMessage request, string host)
+    {
+        // TODO: Use multiple rules, iterating over them until a match is found.
+        var output = _transcoder.Transcode(request);
+        return output.CreateRequest(host);
```
Comment (amanda-tarafa) [inline Google.Api.Gax.Grpc/Rest/RestMethod.cs]: nit: Could these be squashed into one method call `_transcoder.CreateTranscodedRequest(request, how)`? Even if `TranscodingOutput` remains a class (maybe a child class of `HttpRuleTranscoder`).

Or is the separetion useful for testing or something else?

File: Google.Api.Gax.Grpc/Rest/TranscodingOutput.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+using System.Net.Http;
+using System.Text;
+
+namespace Google.Api.Gax.Grpc.Rest;
+
+/// <summary>
+/// The result of transcoding a protobuf request using an HttpRule.
+/// This is produced by <see cref="HttpRuleTranscoder"/>.
+/// </summary>
+internal sealed class TranscodingOutput
```
Comment (amanda-tarafa) [inline Google.Api.Gax.Grpc/Rest/TranscodingOutput.cs:20]: nit: Maybe `TranscodedRequest` and then `CreateRequest` => `CreateHttpRequest`

File: Google.Api.Gax.Grpc/Rest/HttpRulePathPattern.cs [hunk 3]
```diff
@@ -9,158 +9,262 @@
...
 
-        internal string Format(IMessage message) => string.Join("", _segments.Select(segment => segment.Format(message)));
+    /// <summary>
+    /// Attempts to format the path with respect to the given request,
+    /// returning the formatted segment or null if formatting failed due to the request data.
+    /// </summary>
+    internal string TryFormat(IMessage message)
+    {
+        StringBuilder builder = new StringBuilder();
+        var segments = string.Join("", _segments.Select(segment => segment.TryFormat(message)));
```
Comment (amanda-tarafa) [inline Google.Api.Gax.Grpc/Rest/HttpRulePathPattern.cs:81]: I don't think you are using this anymore.

File: Google.Api.Gax.Grpc/Rest/HttpRuleTranscoder.cs [hunk 4]
```diff
@@ -9,89 +9,140 @@
...
+
+            _pathPattern = HttpRulePathPattern.Parse(pattern, requestMessage);
+
+            (bool allFieldsInBody, string bodyName, _bodyTranscoder) = rule.Body switch
+            {
+                // TODO: The body shouldn't contain any fields that are present in the URI.
+                "*" => (true, null, new Func<IMessage, string>(protoRequest => protoRequest.ToString())),
+                "" => (false, null, new Func<IMessage, string>(protoRequest => null)),
+                string name when requestMessage.FindFieldByName(name) is FieldDescriptor field => (false, name, new Func<IMessage, string>(protoRequest => field.Accessor.GetValue(protoRequest).ToString())),
+                _ => throw new ArgumentException($"Method {methodName} has a body parameter {rule.Body} in the 'google.api.http' annotation which is not a field in {requestMessage.Name}")
```
Comment (amanda-tarafa) [inline Google.Api.Gax.Grpc/Rest/HttpRuleTranscoder.cs:83]: Just out of curiosity, we have a presubmit check for this?

File: Google.Api.Gax.Grpc/Rest/HttpRuleTranscoder.cs [hunk 5]
```diff
@@ -9,89 +9,140 @@
...
+            };
+
+            _pathPattern = HttpRulePathPattern.Parse(pattern, requestMessage);
+
+            (bool allFieldsInBody, string bodyName, _bodyTranscoder) = rule.Body switch
+            {
+                // TODO: The body shouldn't contain any fields that are present in the URI.
+                "*" => (true, null, new Func<IMessage, string>(protoRequest => protoRequest.ToString())),
+                "" => (false, null, new Func<IMessage, string>(protoRequest => null)),
+                string name when requestMessage.FindFieldByName(name) is FieldDescriptor field => (false, name, new Func<IMessage, string>(protoRequest => field.Accessor.GetValue(protoRequest).ToString())),
```
Comment (amanda-tarafa) [inline Google.Api.Gax.Grpc/Rest/HttpRuleTranscoder.cs:82]: Can this ever be null?

`field.Accessor.GetValue(protoRequest)?.ToString() ?? ""`

Comment (amanda-tarafa) [reply Google.Api.Gax.Grpc/Rest/HttpRuleTranscoder.cs:82]: Or should that be a reason for failure? It's not clear on the AIP.

File: Google.Api.Gax.Grpc/Rest/HttpRuleTranscoder.cs [hunk 6]
```diff
@@ -9,89 +9,140 @@
...
+                HttpRule.PatternOneofCase.Patch => (rule.Patch, s_patchMethod),
+                HttpRule.PatternOneofCase.Put => (rule.Put, HttpMethod.Put),
+                HttpRule.PatternOneofCase.Delete => (rule.Delete, HttpMethod.Delete),
+                HttpRule.PatternOneofCase.Custom => (rule.Custom.Path, new HttpMethod(rule.Custom.Kind)),
+                _ => throw new ArgumentException("HTTP rule has no pattern")
+            };
+
+            _pathPattern = HttpRulePathPattern.Parse(pattern, requestMessage);
+
+            (bool allFieldsInBody, string bodyName, _bodyTranscoder) = rule.Body switch
```
Comment (amanda-tarafa) [inline Google.Api.Gax.Grpc/Rest/HttpRuleTranscoder.cs:77]: I think that Body being null is valid, seems like it actually cannot be specified for GET and DELETE.

File: Google.Api.Gax.Grpc/Rest/HttpRuleTranscoder.cs [hunk 7]
```diff
@@ -9,89 +9,140 @@
...
+            {
+                // TODO: The body shouldn't contain any fields that are present in the URI.
```
Comment (amanda-tarafa) [inline Google.Api.Gax.Grpc/Rest/HttpRuleTranscoder.cs:79]: What I understand is that the body must not be the same as a URI parameters vs. contain.

I also wonder if it's us that need to check for this. We don't check for many of the other constraints specified in the AIP (as far as I can see, but I might be missing something). For instance, we don't check for:

- The body clause must be identical in the top-level annotation and each additional binding.
- RPCs must not define a body at all for RPCs that use the GET or DELETE HTTP verbs.

If anything these should be presubmit checks on protos?

I also understand that, if we don't have those presubmit checks, and we end up executing requests with bad transcoding, errors that end users get will be confusing. In turn, if we check these constraints ourselves we can at least give meaningful errors. But I guess we either check all or none?

Again, I might be missing something here.

**Review by amanda-tarafa — APPROVED** [review]
LGTM with a few comments/questions possibly because I don't have all the context.

File: Google.Api.Gax.Grpc/Rest/TranscodingOutput.cs (hunk 2 above)
Comment (jskeet) [reply Google.Api.Gax.Grpc/Rest/TranscodingOutput.cs:20]: Maybe later - we tend to only really think about transcoding as the proto->HTTP side of things, so I'm not worried about being ambiguous there. CreateRequest => CreateHttpRequest (or ToHttpRequestMessage) makes sense.

File: Google.Api.Gax.Grpc/Rest/HttpRulePathPattern.cs (hunk 3 above)
Comment (jskeet) [reply Google.Api.Gax.Grpc/Rest/HttpRulePathPattern.cs:81]: Thanks - will fix in an extra tidy-up commit.

File: Google.Api.Gax.Grpc/Rest/RestMethod.cs (hunk 1 above)
Comment (jskeet) [reply Google.Api.Gax.Grpc/Rest/RestMethod.cs]: It's useful for #579 :)

**Comment by jskeet** [comment]
Added an extra commit to address review comments. PTAL.

**Review by amanda-tarafa — APPROVED** [review]
LGTM

**Pull request merged**